Removing classes

noahmanger opened the pull request:
Removes various instances of class names (more details in commit messages) and parametizes the one instance where it makes sense to keep it.

Also opens the accordion with the expand method, rather than faking a click.

This looks bigger than it is because it's checking in the compiled main.js . Do we want to be doing this?

Resolves #3 #5 #7

Commit:
Removing unnecessary classes

- Removes .is-open from the glossary div, since we already have
aria-hidden=false
- Replaces .active on the toggle button with aria-expanded
- Parametizes the highlighted term class

Commit:
Open accordion with expand method

Commit:
Updating examples

Commit:
Select html terms by data attribute instead of class

Resolves #7

Commit:
Updating test

Commit:
Parameterizing classes used by Listjs

Note, Listjs requires class names without the dot

Commit:
Adding a new README

Commit:
More paramaterizing

Commit:
Parameterizing presentational class names

Commit:
Updated readme

noahmanger commented:
Update: I did more to add more additional parameters so that users can pass in an object of presentational class names that will be applied to the glossary term buttons and the divs that contain the definitions. 

However, doing this meant changing the way we use Listjs to initialize. Previously, Listjs populated the DOM for us, but when it does that, it requires the class names of the elements to be the same as the keys in the JSON. 

So instead, I added a new method that populates the DOM and uses whichever classes the user passed in, and then separately initialize Listjs on this existing `<ul>`.

noahmanger commented:
Oh, and I added a bunch of documentation in the readme.

Commit:
Removing apostrophe

msecret commented on example/style.css:
```diff
@@ -22,10 +22,14 @@ body {
 }
 
-.js-accordion [aria-hidden=true] {
+.glossary div[aria-hidden=true] {
```
Do we need the div part of the selector? What about:

`.glossary [aria-hidden=true] { `

msecret commented on example/style.css:
```diff
@@ -22,10 +22,14 @@ body {
   display: none !important;
 }
 
+.glossary button {
```
Is there a chance somebody would have to use an `input` or `a` for this making this selector not work?

noahmanger replied:
Oh, this was just ugly basic styling for the example. Do you think it's still worth re-writing?

msecret replied:
No it's fine.

jmcarp commented on src/glossary.js:
```diff
@@ -89,21 +93,48 @@ function Glossary(terms, selectors) {
+
+/** Initialize list.js list of terms */
+Glossary.prototype.initList = function() {
+  var glossaryId = this.selectors.body.slice(1);
```
These won't work unless the selectors are an ID, a class, and a class, respectively. What if the first selector is a class, or includes attribute selectors, etc?

noahmanger replied:
Well, I specified in the README that those are what they should be. Is there a more fail-proof way that would be better?

jmcarp replied:
Well, if a variable is required to be a class, it seems most reasonable to call it e.g. `listClass` and not `selectors.list`--it's a class, not an arbitrary selector. Or, if we want to keep the current naming, we could throw an error if the user passes in invalid selector values. Better to crash as soon as possible, and for as clear a reason as possible, given bad inputs.

noahmanger replied:
Sounds good.

jmcarp replied:
Looks like we're still `slice`-ing the `selectors.body` selector and assuming it's going to be an ID. I continue to think that we shouldn't call the ID a selector--since it _has_ to be an ID, let's call it `glossaryID`, or something similar. Looks like `list.js` simply calls this parameter `id` (although it actually accepts an ID or a DOM element, meaning our code does the same).

noahmanger replied:
Thanks for catching. I think I just glossed over this one. Updated.

jmcarp commented on src/glossary.js:
```diff
@@ -89,21 +93,48 @@ function Glossary(terms, selectors) {
+    };
+  });
+
+  var itemTemplate = _.template(
```
This template is static--why define it each time this function is called, and not once elsewhere in the file?

noahmanger replied:
:+1:

jmcarp commented on src/glossary.js:
```diff
@@ -89,21 +93,48 @@ function Glossary(terms, selectors) {
+    {interpolate: /\{\{(.+?)\}\}/g}
+  );
+
+  terms.forEach(function(term) {
```
Looks like we're defining a list of objects called `terms`, and then adding them in a `forEach`. How about just the `forEach`:

``` javascript
this.terms.forEach(function(term) {
  var opts = {
    term: term.term,
    definition: term.definition,
    ...
  };
  this.list.insertAdjacentHTML('beforeend', itemTemplate(opts));
});
```

noahmanger replied:
Makes sense.

jmcarp commented on src/glossary.js:
```diff
@@ -89,21 +93,48 @@ function Glossary(terms, selectors) {
+  var self = this;
+  var terms = this.terms.map(function(term) {
+    return {
+      'term': term.term,
```
Not sure that we need this `map`, but in any case, prefer not to use quotes for key names in object declarations.

noahmanger replied:
:+1:

jmcarp commented on src/glossary.js:
```diff
@@ -153,18 +180,16 @@ Glossary.prototype.toggle = function() {
-  this.body.classList.add('is-open');
   this.body.setAttribute('aria-hidden', 'false');
-  this.toggleBtn.classList.add('active');
+  this.toggleBtn.setAttribute('aria-expanded','true');
```
Prefer spaces between arguments. See also line 192.

noahmanger replied:
:+1:

jmcarp commented:
Left a few minor comments. Thanks for cleaning this up @noahmanger!

Commit:
Cleaning up stylistic changes and editing package.json

h/t @jmcarp

Commit:
Adding org name to package.json

Commit:
More readme edits

Commit:
Another update

Commit:
Adding build script

Commit:
Renaming selector properties

h/t @jmcarp

afeld commented on README.md:
```diff
@@ -3,8 +3,98 @@
+    "term": "Glossary",
+    "definition": "A useful tool for finding the definitions of terms"
+  }
+}
```
This should be an array.

noahmanger replied:
Just curious: why's that?

afeld replied:
Take a look at the [`terms.json`](https://github.com/18F/glossary/blob/removing-classes/example/terms.json)...it's using that format. If the outermost thing is an Object (using curly braces), it needs a unique key for each Object within it, e.g.

``` javascript
{
  "Glossary": {
    // ...
  },
  // ...
}
```

Does that make sense?

noahmanger replied:
Oh derp. I just misunderstood what you were suggesting :)

Commit:
Fixing the readme

h/t @afeld

afeld commented:
I don't have a _ton_ of context here, but am excited to use this plugin – I'd advocate for merging this sooner than later so we can do more focused issues and pull requests for changes.

noahmanger commented:
@afeld yes please! I just got back from vacation so sorry for the delay. But if you or @jmcarp think this is ready to merge, go ahead (I just don't want to merge myself).

noahmanger commented:
Ping on this. @afeld or @jmcarp feel free to merge if this looks good.

Commit:
Updating readme

Commit:
Adding glossary item class

Commit:
Fixing tests

Commit:
Fixing class; checking in updated compiled js

Commit:
Adding param for glossary content prefix

jmcarp commented on README.md:
````diff
@@ -3,8 +3,99 @@
+## Download
+### Via npm (coming soon)
+```
+npm install @18f/glossary-panel
````
Just curious--why the `@18f`? Looks like https://www.npmjs.com/package/glossary-panel is available, and it's shorter.

noahmanger replied:
Oops, I misunderstood the guidance on this. Changed it to just `glossary-panel`.

jmcarp commented on README.md:
````diff
@@ -3,8 +3,99 @@
+npm install @18f/glossary-panel
+```
+
+## Create your data
````
There's really no need to make a file, or to call it `terms.json`. Users can pass in an object of terms in the JS.

jmcarp commented on README.md:
````diff
@@ -3,8 +3,99 @@
+```
+
+## Set up your HTML
+The following is the bare-minum HTML needed in your document:
````
Change "bare-minum" to "bare minimum".

jmcarp commented on README.md:
```diff
@@ -3,8 +3,99 @@
+- `Glossary.findTerm(term)`: If the glossary is opens, filters the list down to the term called, expands the term, and highlights the associated term in the DOM
+
+# Styling
+To style the glossary terms and defintions in the accordion list, either use the default classes or whichever ones you passed in. To change the style of the buttons when the accordion elements are expanded, you can select for `aria-expanded="true"]`.
```
Use `[aria-expanded="true"]` or `aria-expanded="true"`

jmcarp commented on README.md:
```diff
@@ -3,8 +3,99 @@
+# Styling
+To style the glossary terms and defintions in the accordion list, either use the default classes or whichever ones you passed in. To change the style of the buttons when the accordion elements are expanded, you can select for `aria-expanded="true"]`.
+
+You will need to add styles for `aria-hidden="true"` in order to hide the glossary panel and the glossary definitions.
```
See above--should this be `[aria-expanded="true"]`?

Commit:
Cleanup

- Renaming package
- Cleaning up README
- Changing parameter name to glossaryID

h/t @jmcarp

Commit:
Merge branch 'master' of https://github.com/18F/glossary into removing-classes

Conflicts:
	package.json

Commit:
Removing branch from accordion dependency

Commit:
changing accordion package name

afeld commented:
I'm sorry I've been so horrible about getting back to you on this. I'm going to merge it now so that we can tackle things in smaller increments in follow-up issues and pull requests.

noahmanger commented:
Thank you @afeld !
